code/datums/traits/_quirk.dm

```diff
@@ -4,120 +4,151 @@
-	var/mob_trait //if applicable, apply and remove this mob trait
+	var/mob_trait //if applicable, apply and remove this mob quirk
 	var/process = FALSE // Does this quirk use on_process()?
-	var/mob/living/quirk_holder
+	var/datum/mind/quirk_holder // The mind that contains this quirk
+	var/mob/living/quirk_target // The mob that will be affected by this quirk
 
-/datum/quirk/New(mob/living/quirk_mob, spawn_effects)
+/datum/quirk/New(datum/mind/quirk_mind, mob/living/quirk_mob, spawn_effects)
 	..()
-	if(!quirk_mob || (human_only && !ishuman(quirk_mob)) || quirk_mob.has_quirk(type))
+	if(!quirk_mind)
 		qdel(src)
 		return
-	quirk_holder = quirk_mob
+	if(!quirk_mob || quirk_mob.has_quirk(type))
+		qdel(src)
+		return
+	quirk_holder = quirk_mind
+	quirk_target = quirk_mob
 	SSquirks.quirk_objects += src
-	to_chat(quirk_holder, gain_text)
-	quirk_holder.roundstart_quirks += src
-	if(mob_trait)
-		ADD_TRAIT(quirk_holder, mob_trait, ROUNDSTART_TRAIT)
+	to_chat(quirk_target, gain_text)
+	quirk_holder.quirks += src
 	if(process)
 		START_PROCESSING(SSquirks, src)
-	RegisterSignal(quirk_holder, COMSIG_PARENT_QDELETING, PROC_REF(handle_parent_del))
+	RegisterSignal(quirk_holder, COMSIG_PARENT_QDELETING, PROC_REF(handle_holder_del))
+	RegisterSignal(quirk_target, COMSIG_PARENT_QDELETING, PROC_REF(handle_mob_del))
+	if(!is_valid_quirk_target(quirk_target)) //at this point the quirk is saved to the mind
+		return
+
+	if(mob_trait)
+		ADD_TRAIT(quirk_target, mob_trait, ROUNDSTART_TRAIT)
 	add()
 	if(spawn_effects)
 		on_spawn()
-		addtimer(CALLBACK(src, PROC_REF(post_add)), 30)
+		addtimer(CALLBACK(src, PROC_REF(post_spawn)), 30)
 
 /datum/quirk/Destroy()
 	if(process)
 		STOP_PROCESSING(SSquirks, src)
 	if(quirk_holder)
 		remove()
 		UnregisterSignal(quirk_holder, COMSIG_PARENT_QDELETING)
-		if(!QDELETED(quirk_holder))
-			to_chat(quirk_holder, lose_text)
-		quirk_holder.roundstart_quirks -= src
+		if(!QDELETED(quirk_target))
+			UnregisterSignal(quirk_target, COMSIG_PARENT_QDELETING)
+			to_chat(quirk_target, lose_text)
+		quirk_holder.quirks -= src
 		if(mob_trait)
-			REMOVE_TRAIT(quirk_holder, mob_trait, ROUNDSTART_TRAIT)
+			REMOVE_TRAIT(quirk_target, mob_trait, ROUNDSTART_TRAIT)
 		quirk_holder = null
+		quirk_target = null
 	SSquirks.quirk_objects -= src
 	return ..()
 
+/* Don't use this, use the mind's transfer_to proc instead */
 /datum/quirk/proc/transfer_mob(mob/living/to_mob)
-	quirk_holder.roundstart_quirks -= src
-	UnregisterSignal(quirk_holder, COMSIG_PARENT_QDELETING)
-	to_mob.roundstart_quirks += src
-	if(mob_trait)
-		REMOVE_TRAIT(quirk_holder, mob_trait, ROUNDSTART_TRAIT)
-		ADD_TRAIT(to_mob, mob_trait, ROUNDSTART_TRAIT)
-	quirk_holder = to_mob
+	UnregisterSignal(quirk_target, COMSIG_PARENT_QDELETING)
+	if(is_valid_quirk_target(quirk_target))
+		if(mob_trait)
+			REMOVE_TRAIT(quirk_target, mob_trait, ROUNDSTART_TRAIT)
+		remove()
+	quirk_target = to_mob
+	if(process)
+		START_PROCESSING(SSquirks, src)
```

Comment:
This doesnt stop processing at any point, if you transfer a mob while keeping the original mob alive issues may arise

Reply:
This shouldn't be an issue since quirk_target is changed on transfer, and from what I've tested none of the processing quirks continue to apply onto the old mob so it should be fine

Reply:
START_PROCESS has a flag check to prevent double processing, this is fine